src/xwrap.rs

```diff
@@ -2,340 +2,248 @@
 
-            let mut x = attrs.x;
-            let mut y = attrs.y;
+        if tree.parent != 0 {
+            let cookie = self
+                .conn
+                .translate_coordinates(tree.parent, tree.root, geometry.x, geometry.y)
+                .ok()?;
+            let coords = cookie.reply().ok()?;
 
-            if parent != 0 {
-                let mut child = 0;
-                xlib::XTranslateCoordinates(
-                    self.handle,
-                    parent,
-                    root,
-                    attrs.x,
-                    attrs.y,
-                    &mut x,
-                    &mut y,
-                    &mut child,
-                );
-            }
+            Some(util::Rect {
+                x: coords.dst_x as i32,
+                y: coords.dst_y as i32,
+                w: geometry.width as i32,
+                h: geometry.height as i32,
+            })
+        } else {
+            Some(util::Rect {
+                x: geometry.x as i32,
+                y: geometry.y as i32,
+                w: geometry.width as i32,
+                h: geometry.height as i32,
+            })
+        }
+    }
 
-            util::Rect {
-                x,
-                y,
-                w: attrs.width,
-                h: attrs.height,
+    fn find_visual(&self, id: xproto::Visualid) -> Option<&xproto::Visualtype> {
+        for screen in &self.conn.setup().roots {
+            for depth in &screen.allowed_depths {
+                for visual in &depth.visuals {
+                    if visual.visual_id == id {
+                        return Some(visual);
+                    }
+                }
             }
         }
+        None
     }
 
-    pub fn get_image(&self, window: xlib::Window, rect: util::Rect) -> Option<Image> {
-        unsafe {
-            let all_planes = !0;
-            let image = xlib::XGetImage(
-                self.handle,
+    pub fn get_image(&self, window: xproto::Window, rect: util::Rect) -> Option<Image> {
+        let all_planes = !0;
+        let cookie = self
+            .conn
+            .get_image(
+                xproto::ImageFormat::Z_PIXMAP,
                 window,
-                rect.x,
-                rect.y,
-                rect.w as std::ffi::c_uint,
-                rect.h as std::ffi::c_uint,
+                rect.x as i16,
+                rect.y as i16,
+                rect.w as u16,
+                rect.h as u16,
                 all_planes,
-                xlib::ZPixmap,
-            );
-
-            Image::from_raw_ximage(image)
-        }
+            )
+            .ok()?;
+        let img = cookie.reply().ok()?;
+
+        let format = *self
+            .conn
+            .setup()
+            .pixmap_formats
+            .iter()
+            .find(|f| f.depth == img.depth)?;
+        let visual = *self.find_visual(img.visual)?;
+        let byte_order = self.conn.setup().image_byte_order;
```

Comment:
Yeah calling setup these many times is odd.

Reply:
Again, it's a simple getter. x11rb itself does this internally, there's no problem with this.